course5/homework04/homework04.cc

```diff
@@ -0,0 +1,105 @@
+#include <cstdio>
+#include <string>
+#include <cstring>
+#include <cassert>
+#include <tuple>
+#include <cerrno>
+
+class PrintError : public std::runtime_error
+{
+public:
+    using std::runtime_error::runtime_error;
+    PrintError() : std::runtime_error(std::strerror(errno))
+    {}
+};
+
+class StreamWriter
+{
+public:
+    StreamWriter(std::string const &filename, std::string const &mode="w")
+    {
+        mFile = std::fopen(filename.c_str(), mode.c_str());
+    }
+
+    StreamWriter(StreamWriter &&rhs)
+    {
+        mFile = rhs.mFile;
+        rhs.mFile = nullptr;
+    }
+
+    StreamWriter(StreamWriter const &) = delete;
+    StreamWriter operator=(StreamWriter const &) = delete;
+
+    ~StreamWriter()
+    {
+        if (mFile)
+            std::fclose(mFile);
+    }
+
+    StreamWriter& operator<<(bool b)
+    {
+        if (std::fprintf(mFile, b ? "true" : "false") < 0) throw PrintError();
+        return *this;
+    }
+
+    StreamWriter& operator<<(int i)
+    {
+        if (std::fprintf(mFile, "%d", i) < 0) throw PrintError();
+        return *this;
+    }
+
+    StreamWriter& operator<<(double d)
+    {
+        if (std::fprintf(mFile, "%lg", d) < 0) throw PrintError();
+        return *this;
+    }
+
+    StreamWriter& operator<<(char const *str)
+    {
+        if (std::fprintf(mFile, "%s", str) < 0) throw PrintError();
+        return *this;
+    }
+
+    StreamWriter& operator<<(std::string const &str)
+    {
+        if (std::fprintf(mFile, "%s", str.c_str()) < 0) throw PrintError();
+        return *this;
+    }
+
+    void flush()
+    {
+        std::fflush(mFile);
+    }
+private:
+    FILE *mFile;
+};
+
+StreamWriter& MyEndl(StreamWriter &writer)
+{
+    writer << "\n";
+    writer.flush();
+    return writer;
+}
+
+StreamWriter& operator<<(StreamWriter &writer, StreamWriter& (*Manipulator)(StreamWriter&))
```

Comment:
Interesting, but rather slower solution.
In most of the cases, compiler won't be able to inline the "manipulator" call.
An it;s not really a "manipulator", because it doesn't/can't manipulate the stream's state.